client/tinychain/ml/optimizer.py

```diff
@@ -1,58 +1,106 @@
 from abc import abstractmethod
+from typing import List, Dict
 
-from tinychain.collection.tensor import Tensor
+from tinychain.collection.tensor import Dense, Tensor
 from tinychain.decorators import closure, post_op
 from tinychain.ref import After, While
 from tinychain.state import Map
 from tinychain.value import F32, UInt
+from tinychain.ml import EPS, Parameter, DiffedParameter
 
 
 class Optimizer(Map):
     @abstractmethod
     def optimize(self, i, gradient, inputs, loss):
         """Update the given `gradient` by computing deltas for the given `loss`."""
 
+    @property
```

Comment:
`self["lr"]` works fine, but type-annotated property enables Python IDE features like completion and type inference. Python language server knows `self.lr` is a float, but it can tell nothing about `self["lr"]`. That's the main motivation for using property vs string indexing.